aten Bool on constant and tensor

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103211

Differential Revision: [D46539695](https://our.internmc.facebook.com/intern/diff/D46539695/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

avikchaudhuri committed:
aten Bool on constant and tensor



Differential Revision: [D46539695](https://our.internmc.facebook.com/intern/diff/D46539695/)

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -528,6 +528,14 @@ def get_state_from_generator():
             assert len(args) == 1, "Expected one arg (pg)"
             assert isinstance(args[0], ProcessGroupVariable)
             return ConstantVariable(self.value(args[0].as_python_constant()))
+        elif self.value == torch.ops.aten.Bool and len(args) == 1:
+            arg = args[0]
+            if isinstance(arg, ConstantVariable):
+                return ConstantVariable(self.value(arg.as_python_constant()))
+            elif isinstance(arg, TensorVariable):
+                if all(isinstance(dim, int) and dim == 1 for dim in arg.size):
+                    return ConstantVariable(self.value(arg.get_real_value().item()))
+            unimplemented(f"{self.value} called on unexpected arg: {arg}")
```
Why are we turning them into constants instead of writing the op into the graph?

anijain2305 replied:
Because it returns a bool (instead of a Tensor)

~~~
>>> import torch
>>> torch.ops.aten.Bool(torch.tensor(5))
True
~~~

But we will end up specializing on this without having a guard, which could lead to incorrect results.

avikchaudhuri replied:
I'm not sure what I'm doing, frankly. Without this intercept we fall back to wrap_fx_proxy(...) which causes the following error:
`torch._dynamo.exc.Unsupported: torch.* op returned non-Tensor ...`

reported in https://github.com/pytorch/pytorch/issues/99752

I guess I'm trying to make it OK to return a constant instead of a tensor here. @yanboliang

yanboliang replied:
Just had a discussion with @anijain2305, my original assessment is not accurate. We can't constant fold it or putting them into FX graph, since this is data dependent op and similar to ```tensor.item()``` call, so the right way is just graph break but we should have better error message.

avikchaudhuri replied:
So is it worth doing this for constant values (not tensor)? Doesn't seem very important, but I can modify this PR if you think it's still worth doing for smaller scope.

avikchaudhuri committed:
Update on "aten Bool on constant and tensor"


Differential Revision: [D46539695](https://our.internmc.facebook.com/intern/diff/D46539695/)

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy

[ghstack-poisoned]
